### PR TITLE
Fix compilation issues with DiskHDFS

### PR DESCRIPTION
**Pull request by zxealous**
Changelog category (leave one):
- Build/Testing/Packaging Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
I test compile ClickHouse without HDFS but failed, because the code in DiskHDFS doesn`t judge whether use HDFS

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Disks/HDFS/DiskHDFS.cpp [hunk 1]
```diff
@@ -1,4 +1,7 @@
 #include <Disks/HDFS/DiskHDFS.h>
+
+#if USE_HDFS
```
Comment (azat) [inline src/Disks/HDFS/DiskHDFS.cpp:3]: But there is https://github.com/ClickHouse/ClickHouse/blob/49f49da06a1fa9605d6771ba887e398506f50147/src/CMakeLists.txt#L118-L121

Why this is not enough?
I guess it has other dependencies except for hdfs library?

Comment (zxealous) [reply src/Disks/HDFS/DiskHDFS.cpp:3]: DiskHDFS.h include Storages/HDFS/HDFSCommon.h, if we don't use hdfs library, and USE_HDFS = 0, code in HDFSCommon will not be compiled, but DiskHDFS use it, so i think DiskHDFS should also add USE_HDFS

Comment (azat) [reply src/Disks/HDFS/DiskHDFS.cpp:3]: Maybe those includes (`#include <>`) should be guarded with `#if USE_HDFS` (and proper inclusion of `Common/config.h`)

BTW can you share origin error?

Comment (zxealous) [reply src/Disks/HDFS/DiskHDFS.cpp:3]: Not just those includes(`#include <>`) , i think maybe all code should be guarded with `#if USE_HDFS ` like my change,   for example, DiskHDFS use the function createHDFSFS which was defined in HDFSCommon, if not USE_HDFS = 0, the function will not be compiled.  

I will copy the error later.

Comment (zxealous) [reply src/Disks/HDFS/DiskHDFS.cpp:3]: `In file included from ../src/Disks/HDFS/DiskHDFS.cpp:1:
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.h:72:5: error: unknown type name 'HDFSBuilderWrapper'
    HDFSBuilderWrapper hdfs_builder;
    ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.h:73:5: error: unknown type name 'HDFSFSPtr'
    HDFSFSPtr hdfs_fs;
    ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:52:80: error: unknown type name 'ReadBufferFromHDFS'; did you mean 'ReadBufferFromFile'?
class ReadIndirectBufferFromHDFS final : public ReadIndirectBufferFromRemoteFS<ReadBufferFromHDFS>
                                                                               ^~~~~~~~~~~~~~~~~~
                                                                               ReadBufferFromFile
/home/ClickHouse/src/IO/ReadBufferFromFile.h:19:7: note: 'ReadBufferFromFile' declared here
class ReadBufferFromFile : public ReadBufferFromFileDescriptor
      ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:69:21: error: unknown type name 'ReadBufferFromHDFS'; did you mean 'ReadBufferFromFile'?
    std::unique_ptr<ReadBufferFromHDFS> createReadBuffer(const String & path) override
                    ^~~~~~~~~~~~~~~~~~
                    ReadBufferFromFile
/home/ClickHouse/src/IO/ReadBufferFromFile.h:19:7: note: 'ReadBufferFromFile' declared here
class ReadBufferFromFile : public ReadBufferFromFileDescriptor
      ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:60:42: error: unknown type name 'ReadBufferFromHDFS'; did you mean 'ReadBufferFromFile'?
        : ReadIndirectBufferFromRemoteFS<ReadBufferFromHDFS>(metadata_)
                                         ^~~~~~~~~~~~~~~~~~
                                         ReadBufferFromFile
/home/ClickHouse/src/IO/ReadBufferFromFile.h:19:7: note: 'ReadBufferFromFile' declared here
class ReadBufferFromFile : public ReadBufferFromFileDescriptor
      ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:71:33: error: unknown type name 'ReadBufferFromHDFS'; did you mean 'ReadBufferFromFile'?
        return std::make_unique<ReadBufferFromHDFS>(hdfs_uri, hdfs_directory + path, config, buf_size);
                                ^~~~~~~~~~~~~~~~~~
                                ReadBufferFromFile
/home/ClickHouse/src/IO/ReadBufferFromFile.h:19:7: note: 'ReadBufferFromFile' declared here
class ReadBufferFromFile : public ReadBufferFromFileDescriptor
      ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:90:20: error: use of undeclared identifier 'createHDFSBuilder'
    , hdfs_builder(createHDFSBuilder(hdfs_root_path_, config))
                   ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:99:21: error: cannot initialize object parameter of type 'const DB::IDiskRemote' with an expression of type 'const DB::DiskHDFS'
    auto metadata = readMeta(path);
                    ^~~~~~~~
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:112:21: error: cannot initialize object parameter of type 'DB::IDiskRemote' with an expression of type 'DB::DiskHDFS'
    auto metadata = readOrCreateMetaForWriting(path, mode);
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:122:41: error: unknown type name 'WriteBufferFromHDFS'; did you mean 'WriteBufferFromFile'?
    auto hdfs_buffer = std::make_unique<WriteBufferFromHDFS>(hdfs_path,
                                        ^~~~~~~~~~~~~~~~~~~
                                        WriteBufferFromFile
/home/ClickHouse/src/IO/WriteBufferFromFile.h:25:7: note: 'WriteBufferFromFile' declared here
class WriteBufferFromFile : public WriteBufferFromFileDescriptor
      ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:126:61: error: use of undeclared identifier 'WriteBufferFromHDFS'
    return std::make_unique<WriteIndirectBufferFromRemoteFS<WriteBufferFromHDFS>>(std::move(hdfs_buffer),
                                                            ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:198:16: error: no viable conversion from returned value of type 'shared_ptr<DB::DiskHDFS>' to function return type 'shared_ptr<DB::IDisk>'
        return std::make_shared<DiskHDFS>(
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ClickHouse/contrib/libcxx/include/memory:2711:23: note: candidate constructor not viable: no known conversion from 'shared_ptr<DB::DiskHDFS>' to 'std::nullptr_t' (aka 'nullptr_t') for 1st argument
    _LIBCPP_CONSTEXPR shared_ptr(nullptr_t) _NOEXCEPT;
                      ^
/home/ClickHouse/contrib/libcxx/include/memory:2725:5: note: candidate constructor not viable: no known conversion from 'shared_ptr<DB::DiskHDFS>' to 'const std::__1::shared_ptr<DB::IDisk> &' for 1st argument
    shared_ptr(const shared_ptr& __r) _NOEXCEPT;
    ^
/home/ClickHouse/contrib/libcxx/include/memory:2732:5: note: candidate constructor not viable: no known conversion from 'shared_ptr<DB::DiskHDFS>' to 'std::__1::shared_ptr<DB::IDisk> &&' for 1st argument
    shared_ptr(shared_ptr&& __r) _NOEXCEPT;
    ^
/home/ClickHouse/contrib/libcxx/include/memory:2728:9: note: candidate template ignored: requirement '__compatible_with<DB::DiskHDFS, DB::IDisk>::value' was not satisfied [with _Yp = DB::DiskHDFS]
        shared_ptr(const shared_ptr<_Yp>& __r,
        ^
/home/ClickHouse/contrib/libcxx/include/memory:2733:52: note: candidate template ignored: requirement '__compatible_with<DB::DiskHDFS, DB::IDisk>::value' was not satisfied [with _Yp = DB::DiskHDFS]
    template<class _Yp> _LIBCPP_INLINE_VISIBILITY  shared_ptr(shared_ptr<_Yp>&& __r,
                                                   ^
/home/ClickHouse/contrib/libcxx/include/memory:2744:9: note: candidate template ignored: could not match 'unique_ptr' against 'shared_ptr'
        shared_ptr(unique_ptr<_Yp, _Dp>&&,
        ^
/home/ClickHouse/contrib/libcxx/include/memory:2753:9: note: candidate template ignored: could not match 'unique_ptr' against 'shared_ptr'
        shared_ptr(unique_ptr<_Yp, _Dp>&&,
        ^
/home/ClickHouse/contrib/libcxx/include/memory:2713:18: note: explicit constructor is not a candidate
        explicit shared_ptr(_Yp* __p,
                 ^
/home/ClickHouse/contrib/libcxx/include/memory:2736:34: note: explicit constructor is not a candidate
    template<class _Yp> explicit shared_ptr(const weak_ptr<_Yp>& __r,
                                 ^
/home/ClickHouse/contrib/libcxx/include/memory:2068:32: error: no matching constructor for initialization of 'DB::ReadBufferFromFile'
    return unique_ptr<_Tp>(new _Tp(_VSTD::forward<_Args>(__args)...));
                               ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:71:21: note: in instantiation of function template specialization 'std::__1::make_unique<DB::ReadBufferFromFile, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, const Poco::Util::AbstractConfiguration &, unsigned long &>' requested here
        return std::make_unique<ReadBufferFromHDFS>(hdfs_uri, hdfs_directory + path, config, buf_size);
                    ^
/home/ClickHouse/src/IO/ReadBufferFromFile.h:26:14: note: candidate constructor not viable: no known conversion from 'std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>' to 'size_t' (aka 'unsigned long') for 2nd argument
    explicit ReadBufferFromFile(const std::string & file_name_, size_t buf_size = DBMS_DEFAULT_BUFFER_SIZE, int flags = -1,
             ^
/home/ClickHouse/src/IO/ReadBufferFromFile.h:30:14: note: candidate constructor not viable: no known conversion from 'std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>' to 'int &' for 1st argument
    explicit ReadBufferFromFile(
             ^
/home/ClickHouse/src/IO/ReadBufferFromFile.h:19:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 4 were provided
class ReadBufferFromFile : public ReadBufferFromFileDescriptor
      ^
In file included from ../src/Disks/HDFS/DiskHDFS.cpp:1:
In file included from ../src/Disks/HDFS/DiskHDFS.h:4:
In file included from ../src/Disks/IDiskRemote.h:8:
In file included from ../src/Disks/DiskFactory.h:3:
In file included from ../src/Disks/IDisk.h:3:
In file included from ../src/Interpreters/Context_fwd.h:3:
In file included from ../src/Common/Exception.h:4:
In file included from ../contrib/libcxx/include/vector:276:
In file included from ../contrib/libcxx/include/__bit_reference:15:
In file included from ../contrib/libcxx/include/algorithm:643:
/home/ClickHouse/contrib/libcxx/include/memory:2068:32: error: no matching constructor for initialization of 'DB::WriteBufferFromFile'
    return unique_ptr<_Tp>(new _Tp(_VSTD::forward<_Args>(__args)...));
                               ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:122:29: note: in instantiation of function template specialization 'std::__1::make_unique<DB::WriteBufferFromFile, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, const Poco::Util::AbstractConfiguration &, unsigned long &, int>' requested here
    auto hdfs_buffer = std::make_unique<WriteBufferFromHDFS>(hdfs_path,
                            ^
/home/ClickHouse/src/IO/WriteBufferFromFile.h:32:5: note: candidate constructor not viable: no known conversion from 'const Poco::Util::AbstractConfiguration' to 'size_t' (aka 'unsigned long') for 2nd argument
    WriteBufferFromFile(
    ^
/home/ClickHouse/src/IO/WriteBufferFromFile.h:41:5: note: candidate constructor not viable: no known conversion from 'std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>' to 'int &' for 1st argument
    WriteBufferFromFile(
    ^
/home/ClickHouse/src/IO/WriteBufferFromFile.h:25:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 4 were provided
class WriteBufferFromFile : public WriteBufferFromFileDescriptor
      ^
In file included from ../src/Disks/HDFS/DiskHDFS.cpp:1:
In file included from ../src/Disks/HDFS/DiskHDFS.h:4:
In file included from ../src/Disks/IDiskRemote.h:8:
In file included from ../src/Disks/DiskFactory.h:3:
In file included from ../src/Disks/IDisk.h:3:
In file included from ../src/Interpreters/Context_fwd.h:3:
In file included from ../src/Common/Exception.h:4:
In file included from ../contrib/libcxx/include/vector:276:
In file included from ../contrib/libcxx/include/__bit_reference:15:
In file included from ../contrib/libcxx/include/algorithm:643:
/home/ClickHouse/contrib/libcxx/include/memory:1349:12: error: static_cast from 'std::__1::__compressed_pair<std::__1::allocator<DB::DiskHDFS>, DB::DiskHDFS> *' to 'std::__1::__compressed_pair<std::__1::allocator<DB::DiskHDFS>, DB::DiskHDFS>::_Base1 *' (aka '__compressed_pair_elem<std::__1::allocator<DB::DiskHDFS>, 0> *'), which are not related by inheritance, is not allowed
    return static_cast<_Base1*>(__pair);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ClickHouse/contrib/libcxx/include/memory:2646:74: note: in instantiation of member function 'std::__1::__compressed_pair<std::__1::allocator<DB::DiskHDFS>, DB::DiskHDFS>::__get_first_base' requested here
            typename _CompressedPair::_Base1* __first = _CompressedPair::__get_first_base(__as_pair);
                                                                         ^
/home/ClickHouse/contrib/libcxx/include/memory:2639:27: note: in instantiation of member function 'std::__1::__shared_ptr_emplace<DB::DiskHDFS, std::__1::allocator<DB::DiskHDFS>>::_Storage::__get_alloc' requested here
            ::new ((void*)__get_alloc()) _Alloc(_VSTD::move(__a));
                          ^
/home/ClickHouse/contrib/libcxx/include/memory:2589:11: note: in instantiation of member function 'std::__1::__shared_ptr_emplace<DB::DiskHDFS, std::__1::allocator<DB::DiskHDFS>>::_Storage::_Storage' requested here
        : __storage_(_VSTD::move(__a))
          ^
/home/ClickHouse/contrib/libcxx/include/memory:3360:55: note: in instantiation of function template specialization 'std::__1::__shared_ptr_emplace<DB::DiskHDFS, std::__1::allocator<DB::DiskHDFS>>::__shared_ptr_emplace<const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::unique_ptr<DB::DiskHDFSSettings, std::__1::default_delete<DB::DiskHDFSSettings>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, const Poco::Util::AbstractConfiguration &>' requested here
    ::new ((void*)_VSTD::addressof(*__guard.__get())) _ControlBlock(__a, _VSTD::forward<_Args>(__args)...);
                                                      ^
/home/ClickHouse/contrib/libcxx/include/memory:3369:19: note: in instantiation of function template specialization 'std::__1::allocate_shared<DB::DiskHDFS, std::__1::allocator<DB::DiskHDFS>, const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::unique_ptr<DB::DiskHDFSSettings, std::__1::default_delete<DB::DiskHDFSSettings>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, const Poco::Util::AbstractConfiguration &, void>' requested here
    return _VSTD::allocate_shared<_Tp>(allocator<_Tp>(), _VSTD::forward<_Args>(__args)...);
                  ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:198:21: note: in instantiation of function template specialization 'std::__1::make_shared<DB::DiskHDFS, const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::unique_ptr<DB::DiskHDFSSettings, std::__1::default_delete<DB::DiskHDFSSettings>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, const Poco::Util::AbstractConfiguration &, void>' requested here
        return std::make_shared<DiskHDFS>(
                    ^
In file included from ../src/Disks/HDFS/DiskHDFS.cpp:1:
In file included from ../src/Disks/HDFS/DiskHDFS.h:4:
In file included from ../src/Disks/IDiskRemote.h:8:
In file included from ../src/Disks/DiskFactory.h:3:
In file included from ../src/Disks/IDisk.h:3:
In file included from ../src/Interpreters/Context_fwd.h:3:
In file included from ../src/Common/Exception.h:4:
In file included from ../contrib/libcxx/include/vector:276:
In file included from ../contrib/libcxx/include/__bit_reference:15:
In file included from ../contrib/libcxx/include/algorithm:643:
/home/ClickHouse/contrib/libcxx/include/memory:1353:12: error: static_cast from 'std::__1::__compressed_pair<std::__1::allocator<DB::DiskHDFS>, DB::DiskHDFS> *' to 'std::__1::__compressed_pair<std::__1::allocator<DB::DiskHDFS>, DB::DiskHDFS>::_Base2 *' (aka '__compressed_pair_elem<DB::DiskHDFS, 1> *'), which are not related by inheritance, is not allowed
    return static_cast<_Base2*>(__pair);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ClickHouse/contrib/libcxx/include/memory:2652:75: note: in instantiation of member function 'std::__1::__compressed_pair<std::__1::allocator<DB::DiskHDFS>, DB::DiskHDFS>::__get_second_base' requested here
            typename _CompressedPair::_Base2* __second = _CompressedPair::__get_second_base(__as_pair);
                                                                          ^
/home/ClickHouse/contrib/libcxx/include/memory:2604:53: note: in instantiation of member function 'std::__1::__shared_ptr_emplace<DB::DiskHDFS, std::__1::allocator<DB::DiskHDFS>>::_Storage::__get_elem' requested here
    _Tp* __get_elem() _NOEXCEPT { return __storage_.__get_elem(); }
                                                    ^
/home/ClickHouse/contrib/libcxx/include/memory:3362:76: note: in instantiation of member function 'std::__1::__shared_ptr_emplace<DB::DiskHDFS, std::__1::allocator<DB::DiskHDFS>>::__get_elem' requested here
    return shared_ptr<_Tp>::__create_with_control_block((*__control_block).__get_elem(), _VSTD::addressof(*__control_block));
                                                                           ^
/home/ClickHouse/contrib/libcxx/include/memory:3369:19: note: in instantiation of function template specialization 'std::__1::allocate_shared<DB::DiskHDFS, std::__1::allocator<DB::DiskHDFS>, const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::unique_ptr<DB::DiskHDFSSettings, std::__1::default_delete<DB::DiskHDFSSettings>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, const Poco::Util::AbstractConfiguration &, void>' requested here
    return _VSTD::allocate_shared<_Tp>(allocator<_Tp>(), _VSTD::forward<_Args>(__args)...);
                  ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.cpp:198:21: note: in instantiation of function template specialization 'std::__1::make_shared<DB::DiskHDFS, const std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, std::__1::unique_ptr<DB::DiskHDFSSettings, std::__1::default_delete<DB::DiskHDFSSettings>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> &, const Poco::Util::AbstractConfiguration &, void>' requested here
        return std::make_shared<DiskHDFS>(
                    ^
In file included from ../src/Disks/HDFS/DiskHDFS.cpp:1:
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.h:72:24: error: private field 'hdfs_builder' is not used [-Werror,-Wunused-private-field]
    HDFSBuilderWrapper hdfs_builder;
                       ^
/home/ClickHouse/src/Disks/HDFS/DiskHDFS.h:73:15: error: private field 'hdfs_fs' is not used [-Werror,-Wunused-private-field]
    HDFSFSPtr hdfs_fs;
              ^
18 errors generated.
[116/453] Building CXX object src/CMakeFiles/dbms.dir/Interpreters/AsynchronousMetrics.cpp.o
ninja: build stopped: subcommand failed.`

**Comment by kitaisreal** [comment]
@Mergifyio update

**Review by kitaisreal — APPROVED** [review]
Generally it is better to explicitly mark code that should be compiled only if some flag is enabled on source level also, not just ignore files in cmake.

**Pull request merged**